databricks-sdk-java/src/main/java/com/databricks/sdk/core/NotebookNativeCredentialsProvider.java

```diff
@@ -0,0 +1,133 @@
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A CredentialsProvider that uses the API token from the command context to authenticate.
+ *
+ * <p>The token and hostname are read from the command context, which can be retrieved through the
+ * dbutils API. As the Java SDK does not depend on DBUtils directly, reflection is used to retrieve
+ * the token. This token should be available wherever the DBUtils API is accessible (i.e. in the
+ * Spark driver).
+ */
+public class NotebookNativeCredentialsProvider implements CredentialsProvider {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(NotebookNativeCredentialsProvider.class);
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  @Override
+  public String authType() {
+    return "runtime";
+  }
+
+  @Override
+  public HeaderFactory configure(DatabricksConfig config) {
+    if (System.getenv("DATABRICKS_RUNTIME_VERSION") == null) {
+      return null;
+    }
+
+    // DBUtils is not available in the Java SDK, so we have to use reflection to get the token.
+    // First, we get the context by calling getContext on the notebook field of dbutils, then we get
+    // the apiKey and
+    // apiUrl fields from the context. If this is successful, we set the host on the config.
+    try {
+      Object dbutils = getDbUtils();
+      Object notebook = getField(dbutils, "notebook");
+      TokenAndUrl testTokenAndUrl = getTokenAndUrl(notebook);
+      if (testTokenAndUrl.url == null) {
+        LOG.debug("Workspace URL is not available, skipping runtime auth");
+      }
+      config.setHost(testTokenAndUrl.url);
+
+      return () -> {
+        Map<String, String> headers = new HashMap<>();
+        TokenAndUrl tokenAndUrl;
+        try {
+          tokenAndUrl = getTokenAndUrl(notebook);
+        } catch (IllegalAccessException
+            | NoSuchMethodException
+            | InvocationTargetException
+            | JsonProcessingException e) {
+          throw new RuntimeException(e);
+        }
+        headers.put("Authorization", String.format("Bearer %s", tokenAndUrl.token));
+        return headers;
+      };
+    } catch (Exception e) {
+      LOG.debug("Failed to get token from dbutils", e);
+      return null;
+    }
+  }
+
+  /** Load the dbutils object initialized by DBR. */
+  private static Object getDbUtils() {
+    try {
+      Class<?> dbutilsHolderClass = Class.forName("com.databricks.dbutils_v1.DBUtilsHolder$");
+      Object dbutilsHolder = dbutilsHolderClass.getDeclaredField("MODULE$").get(null);
+      InheritableThreadLocal<Object> dbutils = getField(dbutilsHolder, "dbutils0");
+      return dbutils.get();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** Reflectively get a field by name from an object. */
+  private static <T> T getField(Object o, String fieldName) {
+    Field f;
```

Comment:
nit: fieldName / Object can be null but given we are only using this in scope of notebook native auth, I think it's okay if we don't have a check.

Reply:
True. I can throw an exception here if needed.